File: FastDTW.py
import numpy as np
from scipy.spatial.distance import euclidean


def DTW_Distance(x, y, radius=1):
    distance, _ = fast_dtw(x, y, radius=radius)
    return distance

def fast_dtw(x, y, radius=1, dist_func=euclidean):
    min_size = radius + 2
    
    if len(x) <= min_size or len(y) <= min_size:
        return dtw(x, y, dist_func)

    x_shrunk = _reduce_by_half(x)
    y_shrunk = _reduce_by_half(y)

    _, low_res_path = fast_dtw(x_shrunk, y_shrunk, radius, dist_func)
    
    projected_path = _expand_path(low_res_path, len(x), len(y))
    
    return _dtw_with_window(x, y, projected_path, radius, dist_func)

def _reduce_by_half(sequence):
    return np.array([
        (sequence[i] + sequence[i + 1]) / 2.0
        for i in range(0, len(sequence) - 1, 2)
    ])

def _expand_path(path, len_x, len_y):
    expanded = set()
    for i, j in path:
        i2, j2 = i * 2, j * 2

        for a in range(max(0, i2 - 1), min(len_x, i2 + 2)):
            for b in range(max(0, j2 - 1), min(len_y, j2 + 2)):
                expanded.add((a, b))
    
    return expanded

def _dtw_with_window(x, y, path, radius, dist_func):
    len_x, len_y = len(x), len(y)
    window = _expand_window(path, len_x, len_y, radius)
    
    cost = np.full((len_x, len_y), np.inf)

    if (0, 0) in window:
        cost[0, 0] = dist_func(x[0], y[0])

    for i in range(len_x):
        for j in range(len_y):
            if (i, j) in window:
                current_dist = dist_func(x[i], y[j])

                if i > 0 and j > 0:
                    min_cost = min(
                        cost[i-1, j], 
                        cost[i, j-1],   
                        cost[i-1, j-1]  
                    )
                    cost[i, j] = current_dist + min_cost
                elif i > 0:
                    cost[i, j] = current_dist + cost[i-1, j]
                elif j > 0:
                    cost[i, j] = current_dist + cost[i, j-1]

    path = _backtrack(cost, window)
    
    return cost[-1, -1], path

def _expand_window(path, len_x, len_y, radius):
    window = set()
    for i, j in path:
        for a in range(max(0, i - radius), min(len_x, i + radius + 1)):
            for b in range(max(0, j - radius), min(len_y, j + radius + 1)):
                window.add((a, b))
    return window

def _backtrack(cost_matrix, window):
    i, j = cost_matrix.shape[0] - 1, cost_matrix.shape[1] - 1
    path = [(i, j)]
    
    while i > 0 or j > 0:
        valid_moves = []
        if i > 0 and (i-1, j) in window:
            valid_moves.append((i-1, j))
        if j > 0 and (i, j-1) in window:
            valid_moves.append((i, j-1))
        if i > 0 and j > 0 and (i-1, j-1) in window:
            valid_moves.append((i-1, j-1))
        
        if not valid_moves:
            break
            
        i, j = min(valid_moves, key=lambda x: cost_matrix[x])
        path.append((i, j))
    
    return path[::-1]

def dtw(x, y, dist_func=euclidean):
    len_x, len_y = len(x), len(y)
    cost = np.full((len_x, len_y), np.inf)
    cost[0, 0] = dist_func(x[0], y[0])
    
    for i in range(1, len_x):
        cost[i, 0] = cost[i-1, 0] + dist_func(x[i], y[0])
    for j in range(1, len_y):
        cost[0, j] = cost[0, j-1] + dist_func(x[0], y[j])
    
    for i in range(1, len_x):
        for j in range(1, len_y):
            current_dist = dist_func(x[i], y[j])
            cost[i, j] = current_dist + min(
                cost[i-1, j],   
                cost[i, j-1],   
                cost[i-1, j-1]   
            )
    
    path = _backtrack(cost, {(i, j) for i in range(len_x) for j in range(len_y)})
    return cost[-1, -1], path

File: FastDTW.java
import java.util.*;

public class FastDTW {

    // Small helper class that lets us store (i, j) in a Set
    // and compare them by CONTENTS, not by array reference.
    private static class IntPair {
        final int i;
        final int j;

        IntPair(int i, int j) {
            this.i = i;
            this.j = j;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            IntPair that = (IntPair) o;
            return i == that.i && j == that.j;
        }

        @Override
        public int hashCode() {
            return Objects.hash(i, j);
        }
    }

    public static double computeFastDTW(double[][] x, double[][] y, int radius) {
        return fastDTW(x, y, radius, FastDTW::euclideanDistance);
    }

    private static double fastDTW(double[][] x, double[][] y, int radius, DistanceFunction distanceFunction) {
        int minSize = radius + 2;

        if (x.length <= minSize || y.length <= minSize) {
            // If sequences are small, just do a normal DTW
            return dtw(x, y, distanceFunction);
        }

        // Otherwise, reduce the size by half
        double[][] xShrunk = _reduceByHalf(x);
        double[][] yShrunk = _reduceByHalf(y);

        // Recursively get a "low resolution" path
        List<int[]> lowResPath = fastDTWPath(xShrunk, yShrunk, radius, distanceFunction);

        // Expand the path
        Set<IntPair> projectedPath = _expandPath(lowResPath, x.length, y.length);

        // Run DTW only in the expanded window
        return _dtwWithWindow(x, y, projectedPath, radius, distanceFunction);
    }

    // Calculate the "path" only, for the shrunk data
    private static List<int[]> fastDTWPath(double[][] x, double[][] y, int radius, DistanceFunction distanceFunction) {
        if (x.length < 2 || y.length < 2) {
            return new ArrayList<>();
        }
        return _dtwWithWindowPath(x, y, radius, distanceFunction);
    }

    private static double[][] _reduceByHalf(double[][] sequence) {
        int reducedSize = (sequence.length + 1) / 2;
        double[][] reduced = new double[reducedSize][sequence[0].length];

        for (int i = 0; i < reducedSize; i++) {
            for (int j = 0; j < sequence[0].length; j++) {
                if (2 * i + 1 < sequence.length) {
                    reduced[i][j] = (sequence[2 * i][j] + sequence[2 * i + 1][j]) / 2.0;
                } else {
                    reduced[i][j] = sequence[2 * i][j];
                }
            }
        }
        return reduced;
    }

    /**
     * Expand the "low resolution" path into a set of points in the high-resolution matrix.
     */
    private static Set<IntPair> _expandPath(List<int[]> path, int lenX, int lenY) {
        Set<IntPair> expanded = new HashSet<>();
        for (int[] pair : path) {
            int i = pair[0] * 2;
            int j = pair[1] * 2;

            // For each node in the shrunk path, add a small neighborhood in the full matrix
            for (int a = Math.max(0, i - 1); a < Math.min(lenX, i + 2); a++) {
                for (int b = Math.max(0, j - 1); b < Math.min(lenY, j + 2); b++) {
                    expanded.add(new IntPair(a, b));
                }
            }
        }
        return expanded;
    }

    /**
     * DTW with a "window" defined by the projected path. Only cells within that window are computed.
     */
    private static double _dtwWithWindow(double[][] x, double[][] y, Set<IntPair> path, int radius, DistanceFunction distanceFunction) {
        int lenX = x.length, lenY = y.length;
        double[][] cost = new double[lenX][lenY];

        // Fill with infinity
        for (double[] row : cost) {
            Arrays.fill(row, Double.POSITIVE_INFINITY);
        }

        // If path set includes (0,0), initialize cost
        if (path.contains(new IntPair(0, 0))) {
            cost[0][0] = distanceFunction.compute(x[0], y[0]);
        }

        for (int i = 0; i < lenX; i++) {
            for (int j = 0; j < lenY; j++) {
                // If (i,j) is not in the window, skip it
                if (!path.contains(new IntPair(i, j))) {
                    continue;
                }

                double currentDist = distanceFunction.compute(x[i], y[j]);

                if (i > 0 && j > 0) {
                    cost[i][j] = currentDist + Math.min(
                            cost[i - 1][j],
                            Math.min(cost[i][j - 1], cost[i - 1][j - 1])
                    );
                } else if (i > 0) {
                    cost[i][j] = currentDist + cost[i - 1][j];
                } else if (j > 0) {
                    cost[i][j] = currentDist + cost[i][j - 1];
                }
            }
        }
        return cost[lenX - 1][lenY - 1];
    }

    /**
     * Simple "full-window" DTW to create a path for smaller sequences. 
     * This returns a path as a List<int[]> (no change needed).
     */
    private static List<int[]> _dtwWithWindowPath(double[][] x, double[][] y, int radius, DistanceFunction distanceFunction) {
        int lenX = x.length, lenY = y.length;
        double[][] cost = new double[lenX][lenY];

        for (double[] row : cost) {
            Arrays.fill(row, Double.POSITIVE_INFINITY);
        }

        // Quick check: radius = 0 means no window, but let's keep it simple
        if (radius == 0) {
            return new ArrayList<>();
        }

        cost[0][0] = distanceFunction.compute(x[0], y[0]);

        for (int i = 1; i < lenX; i++) {
            cost[i][0] = cost[i - 1][0] + distanceFunction.compute(x[i], y[0]);
        }
        for (int j = 1; j < lenY; j++) {
            cost[0][j] = cost[0][j - 1] + distanceFunction.compute(x[0], y[j]);
        }

        for (int i = 1; i < lenX; i++) {
            for (int j = 1; j < lenY; j++) {
                double currentDist = distanceFunction.compute(x[i], y[j]);
                cost[i][j] = currentDist + Math.min(
                        cost[i - 1][j],
                        Math.min(cost[i][j - 1], cost[i - 1][j - 1])
                );
            }
        }

        // Backtrack to build the path
        return _backtrack(cost);
    }

    /**
     * Simple backtrack that returns a List<int[]> representing the alignment path.
     */
    private static List<int[]> _backtrack(double[][] costMatrix) {
        int i = costMatrix.length - 1, j = costMatrix[0].length - 1;
        List<int[]> path = new ArrayList<>();
        path.add(new int[]{i, j});

        while (i > 0 || j > 0) {
            // Potential moves: up, left, diagonal
            List<int[]> moves = new ArrayList<>();
            if (i > 0) moves.add(new int[]{i - 1, j});
            if (j > 0) moves.add(new int[]{i, j - 1});
            if (i > 0 && j > 0) moves.add(new int[]{i - 1, j - 1});

            if (moves.isEmpty()) break;

            // Pick the move with minimum cost
            int[] bestMove = Collections.min(moves, Comparator.comparingDouble(a -> costMatrix[a[0]][a[1]]));

            i = bestMove[0];
            j = bestMove[1];
            path.add(new int[]{i, j});
        }

        Collections.reverse(path);
        return path;
    }

    /**
     * Full DTW (no window). Fallback if sequences are small or radius is large.
     */
    private static double dtw(double[][] x, double[][] y, DistanceFunction distanceFunction) {
        int lenX = x.length, lenY = y.length;
        double[][] cost = new double[lenX][lenY];

        for (double[] row : cost) {
            Arrays.fill(row, Double.POSITIVE_INFINITY);
        }

        cost[0][0] = distanceFunction.compute(x[0], y[0]);

        for (int i = 1; i < lenX; i++) {
            cost[i][0] = cost[i - 1][0] + distanceFunction.compute(x[i], y[0]);
        }
        for (int j = 1; j < lenY; j++) {
            cost[0][j] = cost[0][j - 1] + distanceFunction.compute(x[0], y[j]);
        }

        for (int i = 1; i < lenX; i++) {
            for (int j = 1; j < lenY; j++) {
                double currentDist = distanceFunction.compute(x[i], y[j]);
                cost[i][j] = currentDist + Math.min(
                        cost[i - 1][j],
                        Math.min(cost[i][j - 1], cost[i - 1][j - 1])
                );
            }
        }
        return cost[lenX - 1][lenY - 1];
    }

    private static double euclideanDistance(double[] a, double[] b) {
        double sum = 0;
        for (int i = 0; i < a.length; i++) {
            sum += Math.pow(a[i] - b[i], 2);
        }
        return Math.sqrt(sum);
    }

    @FunctionalInterface
    interface DistanceFunction {
        double compute(double[] a, double[] b);
    }
}
